Detectors/CPV/calib/testWorkflow/GainCalibratorSpec.h

```diff
@@ -0,0 +1,190 @@
+#include "Framework/WorkflowSpec.h"
+#include "Framework/CCDBParamSpec.h"
+#include "CCDB/CcdbApi.h"
+#include "CCDB/CcdbObjectInfo.h"
+#include "DetectorsCalibration/Utils.h"
+#include "CPVCalibration/GainCalibrator.h"
+#include "DataFormatsCPV/Digit.h"
+#include "DataFormatsCPV/TriggerRecord.h"
+#include "DetectorsBase/GRPGeomHelper.h"
+
+using namespace o2::framework;
+
+namespace o2
+{
+namespace calibration
+{
+class CPVGainCalibratorSpec : public o2::framework::Task
+{
+ public:
+  CPVGainCalibratorSpec(std::shared_ptr<o2::base::GRPGeomRequest> req) : mCCDBRequest(req) {}
+
+  //_________________________________________________________________
+  void init(o2::framework::InitContext& ic) final
+  {
+    o2::base::GRPGeomHelper::instance().setRequest(mCCDBRequest);
+    // auto slotL = ic.options().get<uint32_t>("tf-per-slot");
+    // auto delay = ic.options().get<uint32_t>("max-delay");
+    auto updateInterval = ic.options().get<uint32_t>("updateInterval"); // in TF
+    bool updateAtTheEndOfRunOnly = ic.options().get<bool>("updateAtTheEndOfRunOnly");
+    mCalibrator = std::make_unique<o2::cpv::GainCalibrator>();
+    mCalibrator->setSlotLength(0);
+    mCalibrator->setMaxSlotsDelay(1000);
+    if (updateAtTheEndOfRunOnly) {
+      mCalibrator->setUpdateAtTheEndOfRunOnly();
+    }
+    mCalibrator->setCheckIntervalInfiniteSlot(updateInterval);
+    mCalibrator->setUpdateTFInterval(updateInterval);
+    LOG(info) << "CPVGainCalibratorSpec initialized";
+    LOG(info) << "tf-per-slot = 0 (inconfigurable for this calibrator)";
+    LOG(info) << "max-delay = 1000 (inconfigurable for this calibrator)";
+    LOG(info) << "updateInterval = " << updateInterval;
+    LOG(info) << "updateAtTheEndOfRunOnly = " << updateAtTheEndOfRunOnly;
+  }
+
+  void finaliseCCDB(o2::framework::ConcreteDataMatcher& matcher, void* obj) final
+  {
+    o2::base::GRPGeomHelper::instance().finaliseCCDB(matcher, obj);
+  }
+
+  //_________________________________________________________________
+  void run(o2::framework::ProcessingContext& pc) final
+  {
+    o2::base::GRPGeomHelper::instance().checkUpdates(pc);
+    o2::base::TFIDInfoHelper::fillTFIDInfo(pc, mCalibrator->getCurrentTFInfo());
+    TFType tfcounter = mCalibrator->getCurrentTFInfo().startTime;
+    auto&& digits = pc.inputs().get<gsl::span<o2::cpv::Digit>>("calibdigs");
+
+    // previous gains
+    if (!mCalibrator->isSettedPreviousGains()) {
+      const auto previousGains = o2::framework::DataRefUtils::as<CCDBSerialized<o2::cpv::CalibParams>>(pc.inputs().get("gains"));
+      if (previousGains) {
+        mCalibrator->setPreviousGains(new o2::cpv::CalibParams(*previousGains));
+        LOG(info) << "GainCalibratorSpec()::run() : I got previous gains";
+      }
+    }
+
+    // previous calib data
+    if (!mCalibrator->isSettedPreviousGainCalibData()) {
+      const auto previousGainCalibData = o2::framework::DataRefUtils::as<CCDBSerialized<o2::cpv::GainCalibData>>(pc.inputs().get("calibdata"));
```

Comment:
idem

Reply:
Hi @shahor02 
thank you for your comments! I'm implementing as you proposed. 
Question regarding "Fetching ConfigurableParam-based objects": as I understood, it's enought just to fetch ConfigParam object once and do nothing with it: it will be automatically loaded into memory. Is it correct? Then I just use it as before. 
Is it possible to fetch ConfigParams in void init(o2::framework::InitContext& ic) method?

Reply:
Hi @shahor02 
I'm trying to fetch o2::cpv::CPVCalibParams as
```
pc.inputs().get<o2::cpv::CPVCalibParams*>("calibparams");
```
however it does not compile:
```
/home/evdokimov/alice/sw/SOURCES/O2/dev/0/Framework/Core/include/Framework/InputRecord.h: In instantiation of 'decltype(auto) o2::framework::InputRecord::get(R, int) const [with T = o2::cpv::CPVCalibParams*; R = const char*]':
/home/evdokimov/alice/sw/SOURCES/O2/dev/0/Detectors/CPV/calib/testWorkflow/GainCalibratorSpec.h:76:62:   required from here
/home/evdokimov/alice/sw/SOURCES/O2/dev/0/Framework/Core/include/Framework/InputRecord.h:450:46: error: static assertion failed: T is not a supported type
  450 |       static_assert(always_static_assert<T>::value, "T is not a supported type");
      |                                              ^~~~~

```
`o2::framework::DataRefUtils::as<CCDBSerialized<o2::cpv::CPVCalibParams>>(pc.inputs().get("calibparams"));` also does not work. It compiles normally but do not invoke methods responsible for updating singleton class o2::cpv::CPVCalibParams at runtime, just like you said. Do you know what can be a problem?

Reply:
As I wrote here: https://docs.google.com/presentation/d/1WPNixo57zAn_gM6vMCIBOHs-DZqcKAkLX0CwUgO_VYg/edit#slide=id.g12aeb8209f1_3_0, the class should be declared messageable.

Reply:
Hi @shahor02 thanks, managed to get things work (approximately). Now another problem appear:
I pushed to ccdb-test modified CPVCalibParams in which gainDesiredLandauMPV is 201 instead of default 200. I run following code to fetch it:
```
const auto par = pc.inputs().get<o2::cpv::CPVCalibParams*>("calibparams");
      par->printKeyValues(); // print what we get from query
      o2::cpv::CPVCalibParams::Instance().printKeyValues(); // print what we really have as parameters
```
and that is what I get:
```
[2138535:cpv-noise-calibration]: [18:20:47][INFO] Caching in 140547698162288 ptr to CPV/CPV_CalibPars/0 (0x2e1af80)
...
[2138535:cpv-noise-calibration]: CPVCalibParams.gainDesiredLandauMPV : 201		[ CODE ]
...
[2138535:cpv-noise-calibration]: CPVCalibParams.gainDesiredLandauMPV : 200		[ CODE ]

```
Looks like CPVCalibParams is not a singleton?

Reply:
@sevdokim as the page I've referred to says, one should NOT assign the output of the ConfigurableParam fetching. 
It is a singleton by its usage but the since the ROOT persistency requires the class to have a public c-tor, it has it.

Yes, you need to wait for the https://github.com/AliceO2Group/AliceO2/pull/8834 to be merged (or rebase to it).